### PR TITLE
Observers can now join as huggers from eggs and Carriers

### PR DESCRIPTION
**Pull request by ThePiachu**
# About the pull request

Recently we talked about how it would be nice for observers to be able to join as huggers from more than just egg morphers, so figured I'd implement it. Now Carriers function like mobile carriers for the huggers! :D

# Explain why it's good for the game

It makes vanilla Carriers more important and actually incentivises planting eggs, carrying huggers and all that.

# Testing Photographs and Procedure

<details>
<summary>Screenshots & Videos</summary>

https://youtu.be/zNeva3l8gj4

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

</details>

# Changelog

:cl:
add: Observers can now join as Facehuggers using any planted egg or a Carrier.
add: Ghosts now have a verb to join as a Facehugger from a Carrier or an Egg Morpher (egg list would be too cluttered).
code: Refactored the code around joining as a Facehugger to be less Egg Morpher centric.
balance: Rebalanced how many Facehuggers a Hive can support. Now it's based on Xeno population rather than Marine population.
/:cl:

### CODE REVIEW

File: code/modules/mob/dead/observer/observer.dm [hunk 1]
```diff
@@ -973,3 +973,35 @@ This is the proc mobs get to turn into a ghost. Forked from ghostize due to comp
 			&& src.z == nearby_observer.z && get_dist(src, nearby_observer) <= observer_client.view)
 			to_chat(observer_client, SPAN_DEADSAY("<b>[src]</b> points to [A] [nearby_observer.format_jump(A)]"))
 	return TRUE
+
+mob/dead/observer/proc/can_spawn_as_hugger(var/hivenumber)
```
Comment (Watermelon914) [inline code/modules/mob/dead/observer/observer.dm:977]: this needs / at the start and do this for all other definitions

File: code/modules/mob/dead/observer/observer.dm [hunk 2]
```diff
@@ -973,3 +973,35 @@ This is the proc mobs get to turn into a ghost. Forked from ghostize due to comp
...
+	if(alert(usr, "Are you sure you want to become a facehugger?", "Confirmation", "Yes", "No") == "No")
+		return FALSE
+	return TRUE
+
+mob/dead/observer/proc/spawn_as_hugger(atom/A, var/hivenumber)
+	var/mob/living/carbon/Xenomorph/Facehugger/hugger = new /mob/living/carbon/Xenomorph/Facehugger(A.loc, null, hivenumber)
+	usr.mind.transfer_to(hugger, TRUE)
+	hugger.visible_message(SPAN_XENODANGER("A facehugger suddenly emerges out of \the [A]!"), SPAN_XENODANGER("You emerge out of \the [A] and awaken from your slumber. For the Hive!"))
+	playsound(hugger, 'sound/effects/xeno_newlarva.ogg', 25, TRUE)
+	hugger.generate_name()
```
Comment (Watermelon914) [inline code/modules/mob/dead/observer/observer.dm:1007]: Endline is required at the end of a file

File: code/modules/mob/living/carbon/xenomorph/castes/Carrier.dm [hunk 3]
```diff
@@ -301,3 +301,19 @@
...
+		to_chat(user, SPAN_WARNING("\The [src] doesn't have any facehuggers to inhabit."))
+		return
+	if(!user.can_spawn_as_hugger(hivenumber))
+		return
+	//Need to check again because time passed due to the confirmation window
+	if(!huggers_cur)
+		to_chat(user, SPAN_WARNING("\The [src] doesn't have any facehuggers to inhabit."))
+		return
+	user.spawn_as_hugger(src, hivenumber)
+	huggers_cur--
```
Comment (Watermelon914) [inline code/modules/mob/living/carbon/xenomorph/castes/Carrier.dm:319]: Endline is required at the end of a file

File: code/modules/mob/living/carbon/xenomorph/xeno_defines.dm [hunk 4]
```diff
@@ -264,6 +264,12 @@
 
 	var/need_round_end_check = FALSE
 
+	//Joining as Facehugger vars
+	var/hugger_timelock = 15 MINUTES
+	var/last_marine_count = -5 MINUTES
+	var/marine_count_cooldown = 2 MINUTES
+	var/playable_hugger_limit = 4
```
Comment (Watermelon914) [inline code/modules/mob/living/carbon/xenomorph/xeno_defines.dm:271]: These need doc comments

**Review by Watermelon914 — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/xenomorph/xeno_defines.dm [hunk 5]
```diff
@@ -865,6 +869,45 @@
 		return FALSE
 	return TRUE
 
+/datum/hive_status/proc/update_hugger_limit()
+	playable_hugger_limit = 2 + Ceiling(totalXenos.len / 8)
```
Comment (stanalbatross) [inline code/modules/mob/living/carbon/xenomorph/xeno_defines.dm:873]: this value is pretty harsh, I would drop it down to / 4 or something

File: code/modules/mob/living/carbon/xenomorph/xeno_defines.dm [hunk 6]
```diff
@@ -865,6 +869,45 @@
...
+	if(!GLOB.hive_datum || ! GLOB.hive_datum[hivenumber])
+		return
+	if(world.time < hugger_timelock)
+		to_chat(user, SPAN_WARNING("The hive cannot support facehuggers yet..."))
+		return FALSE
+	if(world.time - user.timeofdeath < 3 MINUTES)
+		var/time_left = round((user.timeofdeath + 3 MINUTES - world.time) / 10)
+		to_chat(user, SPAN_WARNING("You ghosted too recently. You cannot become a facehugger until 3 minutes have passed ([time_left] seconds remaining)."))
+		return FALSE
+	if(totalXenos.len == 0)
```
Comment (stanalbatross) [inline code/modules/mob/living/carbon/xenomorph/xeno_defines.dm:885]:
```suggestion
	if(totalXenos.len <= 0)
```

File: code/modules/mob/living/carbon/xenomorph/xeno_defines.dm [hunk 7]
```diff
@@ -865,6 +869,45 @@
...
+/datum/hive_status/proc/update_hugger_limit()
+	playable_hugger_limit = 2 + Ceiling(totalXenos.len / 8)
+
+/datum/hive_status/proc/can_spawn_as_hugger(mob/dead/observer/user)
+	if(!GLOB.hive_datum || ! GLOB.hive_datum[hivenumber])
+		return
+	if(world.time < hugger_timelock)
+		to_chat(user, SPAN_WARNING("The hive cannot support facehuggers yet..."))
+		return FALSE
+	if(world.time - user.timeofdeath < 3 MINUTES)
```
Comment (stanalbatross) [inline code/modules/mob/living/carbon/xenomorph/xeno_defines.dm:881]: use define for 3 minutes

**Review by stanalbatross — CHANGES_REQUESTED** [review]

**Comment by ThePiachu** [comment]
Added a verb for ghosts to join as a Facehugger, it lists only Carriers and Morphers that have available huggers in them
image

**Comment by ThePiachu** [comment]
You can now set reserves for egg morpher as well if you're a Drone caste.

File: code/modules/mob/living/carbon/xenomorph/abilities/carrier/carrier_abilities.dm [hunk 8]
```diff
@@ -19,3 +19,7 @@
 	macro_path = /datum/action/xeno_action/verb/verb_retrieve_egg
 	action_type = XENO_ACTION_CLICK
 	ability_primacy = XENO_PRIMARY_ACTION_4
+
+/datum/action/xeno_action/onclick/set_hugger_reserve
+	name = "Set Hugger Reserve"
+	action_icon_state = "xeno_banish"
```
Comment (harryob) [inline code/modules/mob/living/carbon/xenomorph/abilities/carrier/carrier_abilities.dm:25]: eol - editorconfig vsc plugin does this for you

File: code/modules/mob/living/carbon/xenomorph/xeno_defines.dm [hunk 9]
```diff
@@ -264,6 +265,10 @@
 
 	var/need_round_end_check = FALSE
 
+	//Joining as Facehugger vars
+	var/hugger_timelock = 15 MINUTES //You can only join as a hugger 15 minutes into the round
+	var/playable_hugger_limit = 0 //How many huggers can the hive support
```
Comment (harryob) [inline code/modules/mob/living/carbon/xenomorph/xeno_defines.dm:270]:
```suggestion
	/// When can huggers join the round
	var/hugger_timelock = 15 MINUTES
	 /// How many huggers can the hive support
	var/playable_hugger_limit = 0
```
i think walter meant to dm doc these vars - so they show up on mouseover in vsc, but this is optional regardless

**Review by harryob — CHANGES_REQUESTED** [review]

**Review by harryob — APPROVED** [review]

**Pull request merged**